### PR TITLE
[NFC] Add a separate cache for nominal signature types

### PR DESCRIPTION
**Pull request by tlively**
We have always cached nominal signature types keyed on their signatures to avoid
creating extra nominal types through the `HeapType::HeapType(Signature)`
constructor. However, that logic was previously built into the HeapTypeInfo
canonicalization system. To allow that system to be simplified in future PRs,
separate the caching into its own explicit layer.

### CODE REVIEW

File: src/wasm/wasm-type.cpp [hunk 1]
```diff
@@ -3241,15 +3263,24 @@ std::vector<HeapType> TypeBuilder::build() {
...
   }
-  return result;
+
+  // Note built signature types. See comment in `HeapType::HeapType(Signature)`.
+  for (auto type : results) {
+    if (type.isSignature() &&
+        (type.isNominal() || getTypeSystem() == TypeSystem::Nominal)) {
+      nominalSignatureCache.insertType(type);
+    }
+  }
```
Comment (kripken) [inline src/wasm/wasm-type.cpp:3281]: Why is this needed when we have it handled in the constructor of HeapType?

Comment (tlively) [reply src/wasm/wasm-type.cpp:3281]: `TypeBuilder` only uses the lower-level constructor that takes a `uintptr_t` pointing to an underlying `HeapTypeInfo`, so that logic in the other constructor does not otherwise get a chance to run.

Comment (kripken) [reply src/wasm/wasm-type.cpp:3281]: Thanks, right, that makes sense...

**Review by kripken — APPROVED** [review]
